### PR TITLE
DM-38952: add new middleware CI package

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: ups/ci_middleware.table [hunk 1]
```diff
@@ -1,9 +1,10 @@
-# List EUPS dependencies of this package here.
-# - Any package whose API is used directly should be listed explicitly.
-# - Common third-party packages can be assumed to be recursively included by
-#   the "base" package.
-setupRequired(base)
+setupRequired(sconsUtils)
```
Comment (timj) [inline ups/ci_middleware.table:1]: We really need sphgeom in this list as well since it's used directly by the code here.

**Review by timj — APPROVED** [review]

File: ups/ci_middleware.table (hunk 1 above)
Comment (timj) [reply ups/ci_middleware.table:1]: Also `resources`, `afw`, `daf_base`, and  `geom` are used.

File: ups/ci_middleware.table [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
+setupRequired(daf_butler)
+setupRequired(pipe_base)
+setupRequired(obs_base)
```
Comment (timj) [inline ups/ci_middleware.table:4]: This doesn't seem to be used explicitly.

File: doc/changes/README.rst [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+Each file should be named like ``<JIRA TICKET>.<TYPE>`` with a file extension defining the markup format.
+The ``<TYPE>`` should be one of:
+
+* ``feature``: New feature
+* ``bugfix``: A bug fix.
+* ``api``: An API change.
+* ``perf``: A performance enhancement.
+* ``doc``: A documentation improvement.
+* ``removal``: An API removal or deprecation.
+* ``other``: Other Changes and Additions of interest to general users.
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply doc/changes/README.rst:16]: I believe I lifted the package configuration files from `obs_base`, and indeed its `doc/changes/README.rst` does still reference "other" as well.  I've fixed it here, but will leave `obs_base` for another ticket.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+-------------
+
+When building QuantumGraphs and running them via `data/SConscript`, the `pipetask` tool is configured to write coverage information to `.coverage*` files in the root directory of the package.
+These can be combined with coverage information from the unit tests for various middleware packages in order to generate a complete coverage report using [`coverage combine`](https://coverage.readthedocs.io/en/7.2.6/cmd.html#combining-data-files-coverage-combine).
+
+Parallelization and Timing
+--------------------------
+
+The SCons build tree defined by `data/SConscript` looks more or less like three independent serial chains (alternating QuantumGraph generation and QuantumBackedButler execution for each of the three pipelines), with independent direct-execution steps branching off after each QuantumGraph generation step.
+As a result, running SCons with more than `-j4` essentially does not help at all, with the minimum time to execute all test runs and unit test scripts around 20 minutes.
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply README.md:68]: Other than the benchmarks with different `-j` and some efforts to shrink the test dataset, I haven't worked on performance much.  I agree a profile would be very interesting, but I'd like to defer it to another ticket.  I'm not sure how best to do that given that there are many processes launched (even when there is no parallelization) and I haven't yet spotted a way to combine Python profiler output files.

[the remark this answers is not in the document]
Comment (timj) [reply README.md:68]: I wasn't asking for the profile on this ticket. `pytest --profile --profile-svg` handles multiple profiles just fine (and you get the individual profiles plus the combined one) so it seems possible.

**Pull request merged**